test/integration/specs/kbaseNarrativeDataPanel_test.js

```diff
@@ -0,0 +1,180 @@
+}
+
+async function selectTab(slideoutPanel, name) {
+    const tab = await slideoutPanel.$(`[data-test-id="tab-${name}"]`);
+    await tab.waitForExist();
+    await clickWhenReady(tab);
+    // wait for tab to have the active class.
+    await waitForClass(tab, 'active');
+
+
+
+    const tabPanel = await slideoutPanel.$(`[data-test-id="panel-${name}"]`);
+    await tabPanel.waitForExist();
+    await waitForClass(tabPanel, 'active');
+    // and wait for it to have the active class.
+
+    return [tab, tabPanel]
+}
+
+async function selectTabByLabel(slideoutPanel, name) {
+    const tab = await slideoutPanel.$(`[data-test-id="tab-${name}"]`);
+    const tabLabel = await tab.$(`[data-test-id="label"]`);
+    await tab.waitForExist();
+    await clickWhenReady(tabLabel);
+    // wait for tab to have the active class.
+    await waitForClass(tab, 'active');
+
+
+    const tabPanel = await slideoutPanel.$(`[data-test-id="panel-${name}"]`);
+    await tabPanel.waitForExist();
+    await waitForClass(tabPanel, 'active');
+    // and wait for it to have the active class.
+
+    return [tab, tabPanel]
+}
+
+
+describe('Tabbing within the data panel should work', () => {
+    beforeEach(async () => {
+        await browser.setTimeout({ 'implicit': 30000 });
+        await browser.reloadSession();
+    });
+
+    afterEach(async () => {
+        await browser.deleteCookies();
+    });
+
+    it('opens the "Shared With Me" tab', async () => {
+        const testCase = testCases.TEST_CASE_1;
+        await login();
+        await openNarrative(testCase.narrativeId);
+        
+        const slideoutPanel = await openDataSlideout();
+
+        const [tab, tabPanel] = await selectTab(slideoutPanel, 'sharedwithme');
+
+        // TODO: make sure the panel content displayed correctly
+    });
+
+    it('opens the "Public" tab', async () => {
+        const testCase = testCases.TEST_CASE_1;
+        await login();
+        await openNarrative(testCase.narrativeId);
+        
+        const slideoutPanel = await openDataSlideout();
+
+        const [tab, tabPanel] = await selectTab(slideoutPanel, 'public');
+        // TODO: make sure the panel content displayed correctly
+    });
+
+    it('opens the "Example" tab', async () => {
+        const testCase = testCases.TEST_CASE_1;
+        await login();
+        await openNarrative(testCase.narrativeId);
+        
+        const slideoutPanel = await openDataSlideout();
+
+        const [tab, tabPanel] = await selectTab(slideoutPanel, 'example');
+        // TODO: make sure the panel content displayed correctly
+    }); 
+
+    it('opens the "Import" tab', async () => {
+        const testCase = testCases.TEST_CASE_1;
+        await login();
+        await openNarrative(testCase.narrativeId);
+
+        const slideoutPanel = await openDataSlideout();
+
+        const [tab, tabPanel] = await selectTab(slideoutPanel, 'import');
+        // TODO: make sure the panel content displayed correctly
+    });
```

Comment:
Rather than repeating very similar code blocks, can you refactor this code so that you iterate through an array (or object, with the tab names as values) of tabs, running the `selectTab` and `selectTabByLabel` tests for each?

Reply:
I think the thing here is that it starts each test case from scratch.
Though I suppose the first several lines could be factored into their own little accessory function.

Reply:
Yup -- I meant that each test case would start from scratch, i.e. running the login and open narrative code and then do the specific `selectTab` or `selectTabByLabel` function.

Reply:
There are upsides and downsides to table-based testing like that. I often love doing it, because you can cram a lot of test cases into fewer lines, but the downside is that you lose information in test failures without extra work. And they do test different things in this case. The isolated tests each open the narrative, open the data slideout, and navigate to a tab, which is different than opening a narrative and navigating amongst the tabs. 

These tests provide easy location of which tab fails if a bug is introduced to one of them.

There is a later test which does repeated bashing of tab selection. It is not ideal because it randomizes the tabs, making the test unreproducible, with the benefit of being accomplished in just a few lines of code. I guess that test could be frozen by generating 100 or whatever tab ids and storing in a json file. It is just a matter of that gift we all possess in very limited supply ... time.

Yeah, the 5 line tests could be bundled up into a parameterized test function. Worth? 

There are so many test cases lurking the the possible combinations in actions provided by tab switching and interaction with other narrative elements, and so many attributes one could test, it is worth considering limiting our expenditure to achieve the goal of ensuring that the major usage cases of what we are testing are covered. In this case, I think it is primarily tab operation with clicking on the tab itself, and clicking on the tab label.

We should continue to improve and add tests to cover the myriad usage cases, but, gosh, I'm just happy to have tests the provide decent coverage of features and are largely readable.

Reply:
In case it wasn't completely clear, this is what I am suggesting:

```js
  const tabNames = {
    import: 'Import',
    sharedwithme: 'Shared With Me',
    ...
  }

  Object.keys(tabNames).forEach((tabId) => {
    it(`opens the "${tabName[tabId]}" tab`, async () => {
         const testCase = testCases.TEST_CASE_1;
         await login();
         await openNarrative(testCase.narrativeId);

         const slideoutPanel = await openDataSlideout();
         const [tab, tabPanel] = await selectTab(slideoutPanel, tabId);
         // TODO: make sure the panel content displayed correctly
    });
  });

Reply:
Oh, you mean to hoist the array of tab ids into a top level variable. Sure, added to the test data.
Also have added some additional tests.